k9mail/src/main/java/com/fsck/k9/controller/QresyncSyncInteractor.java

```diff
@@ -0,0 +1,128 @@
+package com.fsck.k9.controller;
+
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.fsck.k9.Account;
+import com.fsck.k9.mail.Flag;
+import com.fsck.k9.mail.Message;
+import com.fsck.k9.mail.MessagingException;
+import com.fsck.k9.mail.store.imap.ImapFolder;
+import com.fsck.k9.mail.store.imap.ImapMessage;
+import com.fsck.k9.mail.store.imap.QresyncParamResponse;
+import com.fsck.k9.mailstore.LocalFolder;
+import timber.log.Timber;
+
+
+class QresyncSyncInteractor {
+
+    private final Account account;
+    private final LocalFolder localFolder;
+    private final ImapFolder imapFolder;
+    private final MessagingListener listener;
+    private final MessagingController controller;
+    private final ImapSyncInteractor imapSyncInteractor;
+
+    QresyncSyncInteractor(Account account, LocalFolder localFolder, ImapFolder imapFolder, MessagingListener listener,
+            MessagingController controller, ImapSyncInteractor imapSyncInteractor) {
+        this.account = account;
+        this.localFolder = localFolder;
+        this.imapFolder = imapFolder;
+        this.listener = listener;
+        this.controller = controller;
+        this.imapSyncInteractor = imapSyncInteractor;
+    }
+
+    int performSync(QresyncParamResponse qresyncParamResponse, List<String> expungedUids, MessageDownloader messageDownloader)
+            throws MessagingException, IOException {
+        final List<ImapMessage> remoteMessages = new ArrayList<>();
+
+        if (account.syncRemoteDeletions()) {
+            List<String> deletedUids = new ArrayList<>(expungedUids);
+            deletedUids.addAll(qresyncParamResponse.getExpungedUids());
+            imapSyncInteractor.syncRemoteDeletions(deletedUids);
+        }
+
+        findNewRemoteMessagesToDownload(remoteMessages, qresyncParamResponse, messageDownloader);
+        if (imapFolder.getMessageCount() >= localFolder.getVisibleLimit() && imapFolder.getMessageCount() >=
+                (remoteMessages.size() + localFolder.getMessageCount())) {
+            findOldRemoteMessagesToDownload(remoteMessages);
+        }
+
+        return messageDownloader.downloadMessages(account, imapFolder, localFolder, remoteMessages, false, true, false);
+    }
+
+    private void findNewRemoteMessagesToDownload(List<ImapMessage> remoteMessages, QresyncParamResponse qresyncParamResponse,
```

Comment:
`remoteMessages` also isn't a very descriptive name, took me a while to understand what it's used for. `remoteMessageDownloadQueue` maybe?